### PR TITLE
Dashboard Scene:  Improve Angular experience

### LINKED ISSUE #90359: Scenes: Improve Angular experience
## Context
When rendering angular panels in Scenes:
- They don't break. They are rendered with no warning
- The options are not rendered when editing a panel
- The migration is not done or you cannot trigger the migration

This is problematic because:
- We don't offer a way to migrate
- We don't warn users, we just treat Angular panels as React panels

## How to solve it
Let's try to port the maximum capabilities from old architecture to keep the Angular migration as less disruptive as possible

- [ ] Display a warning in the panel header when rendering an angular panel
- [ ] Display a warning in the dashboard when rendering one or more angular panels
- [ ] Display a button to migrate in the dashboard banner 
- [ ] Display a button to migrate in the panel editor
- [ ] Display a button to "Edit options" when it is an Angular panel, to open the panel JSON inspector to be able to edit the options

### PR DESCRIPTION
**Pull request by axelavargas**
**What is this feature?**

When rendering angular panels in Scenes:

* They don't break.  However, they are rendered with no warning
* The options are not rendered when editing a panel
* The migration is not done or you cannot trigger the migration

https://github.com/user-attachments/assets/6b80b994-2fd0-442e-a674-01df22fcd484

In this PR, I have implemented:

- [x] Display a warning in the panel header when rendering an angular panel
- [x] Display a warning in the dashboard when rendering one or more angular panels
- [x] Display a button to migrate in the dashboard banner
- [x] Display a button to migrate in the panel editor
- [x] Display a button to "Edit options" when it is an Angular panel, to open the panel JSON inspector to be able to edit the options
- [x] Support library panels
- [x] Fix TS Errors
- [x] Add tests

#### Limitation 
The `Apply` button from the panel inspect JSON does not work. We have a bug in the Scene migration. This bug will be addressed on a different PR.

**Why do we need this feature?**
We don't offer a way to migrate
We don't warn users, we just treat Angular panels as React panels

**Who is this feature for?**
Users that still have dashboards that contain angular panels

**Which issue(s) does this PR fix?**:

Fixes #90359

**Special notes for your reviewer - To test this PR:**
Make sure you have in your custom.ini or grafana.ini this configuration

```
angular_support_enabled = true

[feature_toggles]
enable = angularDeprecationUI

```
Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

File: public/app/features/dashboard-scene/scene/DashboardScene.tsx [hunk 1]
```diff
@@ -884,6 +885,41 @@ export class DashboardScene extends SceneObjectBase<DashboardSceneState> {
     });
   }
 
+  public getDashboardPanels = () => {
```
Comment (torkelo) [inline public/app/features/dashboard-scene/scene/DashboardScene.tsx:888]: this looks like it duplicates dashboardSceneGraph.getVizPanels

File: public/app/features/dashboard-scene/scene/DashboardScene.tsx [hunk 2]
```diff
@@ -884,6 +885,41 @@ export class DashboardScene extends SceneObjectBase<DashboardSceneState> {
...
+    }
+    return gridItems.map((vizPanel) => {
+      if (!(vizPanel instanceof DashboardGridItem)) {
+        throw new Error('Trying to get a panel in a layout that is not DashboardGridItem');
+      }
+      return vizPanel.state.body;
+    });
+  };
+
+  public hasDashboardAngularPlugins = () => {
```
Comment (torkelo) [inline public/app/features/dashboard-scene/scene/DashboardScene.tsx:906]: why an instance / arrow function? Always try to use prototype functions as a default

File: public/app/features/dashboard-scene/scene/AngularDeprecation.tsx [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    return 'panel';
+  }
+  if (isAngularDatasource) {
+    return 'data source';
+  }
+  return 'panel or data source';
+}
+
+export function isGridPanelAngularPlugin(gridItem: SceneGridItemLike) {
+  //check if panel is library panel
```
Comment (torkelo) [inline public/app/features/dashboard-scene/scene/AngularDeprecation.tsx:73]: is this comment a todo?

File: public/app/features/dashboard-scene/scene/AngularDeprecation.tsx [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  if (isAngularPanel) {
+    return 'panel';
+  }
+  if (isAngularDatasource) {
+    return 'data source';
+  }
+  return 'panel or data source';
+}
+
+export function isGridPanelAngularPlugin(gridItem: SceneGridItemLike) {
```
Comment (torkelo) [inline public/app/features/dashboard-scene/scene/AngularDeprecation.tsx:72]: this parameter looks wronglytyped and named,  should it not take in a VizPanel?

File: public/app/features/dashboard-scene/scene/AngularDeprecation.tsx [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    return (
+      (config.panels[gridItem.state.pluginId]?.angular?.detected ||
+        explicitlyControlledMigrationPanels.includes(gridItem.state.pluginId)) &&
+      !config.panels[gridItem.state.pluginId]?.angular?.hideDeprecation
+    );
+  }
+  return false;
+}
+
+export function isGridPanelAngularDatasource(gridItem: SceneGridItemLike) {
```
Comment (torkelo) [inline public/app/features/dashboard-scene/scene/AngularDeprecation.tsx:84]:
```suggestion
export function isUsingAngularPlugin(panel: VizPanel) {
```

File: public/app/features/dashboard-scene/scene/DashboardSceneRenderer.tsx [hunk 6]
```diff
@@ -55,6 +57,31 @@ export function DashboardSceneRenderer({ model }: SceneComponentProps<DashboardS
     <DashboardEmpty dashboard={model} canCreate={!!model.state.meta.canEdit} key="dashboard-empty-state" />
   );
 
+  const showAngularDeprecationNotice = () => {
```
Comment (torkelo) [inline public/app/features/dashboard-scene/scene/DashboardSceneRenderer.tsx:60]: would turn this into a plain react component named `DashboardAngularDeprecationBanner` or something 

alternatively rename it to module level function named `renderAngularDeprecationBanner(model) (outside this react components render)

nitpick: but function name showXX does not feel like react render function but more like a state action

**Comment by torkelo** [comment]
Looks good! left some comments, mostly nitpicks / suggestions. Did not give a test drive though

File: public/app/features/dashboard-scene/scene/DashboardScene.tsx (hunk 1 above)
Comment (axelavargas) [reply public/app/features/dashboard-scene/scene/DashboardScene.tsx:888]: ahh nice, I will use this one instead

File: public/app/features/dashboard-scene/scene/AngularDeprecation.tsx (hunk 3 above)
Comment (axelavargas) [reply public/app/features/dashboard-scene/scene/AngularDeprecation.tsx:73]: Nope, was a left over, after the library panel simplification PR this is not needed anymore :)

File: public/app/features/dashboard-scene/scene/DashboardScene.tsx (hunk 2 above)
Comment (axelavargas) [reply public/app/features/dashboard-scene/scene/DashboardScene.tsx:906]: no reason behind this, tbh muscle memory without much thinking. I will keep this in mind, thanks for the suggestion

**Review by torkelo — APPROVED** [review]
Code wise it looks good, will take it for a short spin in 1 sec (building)

**Review by ivanortegaalba — APPROVED** [review]
Wow! Thanks Alexa! Works like a charm. This exposes the known bug for apply, but let's open a separate PR for that. This already helps a lot showing the angular warning for non supported charts. That helps a lot! Thank you Alexa!

✅ The warning shows for migratable and non migratable panels
✅ The options are properly migrated (tested with Graph)
✅ The "Try migrate" works at dashboard and panel level

**Pull request merged**